### PR TITLE
Assignment 4 With @GeorgeKoralski007

### PR DESCRIPTION
**Pull request by crawfordkid2**

### CODE REVIEW

**Comment by slanante** [comment]
Will review.

**Comment by BenagainB** [comment]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ✔
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.
Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔
Extra Credit
Refactor your project files to put common properties and/or items into a single file. ✔
Convert ILogger.CreateLogger to a generic method. ✔
Returns the logger classes type that is to be created. ✔
Has a type parameter for the configuration type ✔

Man, I really think you nailed this one! Only changes I'd suggest would be to go back through your files and see what comments can just be ripped out (FullNameRecord.cs is absolutely packed with comments). Other than that, it looks excellent and I may have to go back now and make sure my generic method is written as cleanly as yours.

File: Logger/FullNameRecord.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    //this is the record that turns someones full name into a name, (First last middle) = name using our abstract class
+    public record Someone(FullName Full): AbstractEntity
+    {
+        public override string Name { get => Full.ToString(); }
+    }
+    //the employee record uses fullname and the someone record to assign a full name and make them someone implicitly
+    public record Employee(FullName Full, string Position, string Company) : Someone(Full);
+    //the student record is very similar to the employee but is slightly different with what fields are needed.
+    public record Student(FullName Full, string Degree, string School) : Someone(Full);
+}
```
Comment (slanante) [inline Logger/FullNameRecord.cs:42]: I would recommend placing these classes in their own files. The FullNameRecord file does not have a name that would suggest the Emplyoee, Student, and Book classes are contained within. I would also suggest placing the Someone class inside its own class file as well.

Comment (slanante) [reply Logger/FullNameRecord.cs:42]: You are also not setting attributes for the Book, Employee or Student class. You are passing parameters in Book for example, but the only parameter that is saved by a book object is Name. You have not set Author or ISBN as attributes of Book and these parameters will not be saved into an instance of Book. Equality checks will fail for your current implementation of these classes but with a correct implementation they should pass. Below is a screenshot of a test I added on to your code. 
image

**Review by slanante — COMMENTED** [review]
Overall, good job. There is a need to edit the Book, Employee and Student records as their checks for equality will fail with the current implementation. Separating these same classes into their own files will also help improve the organization and help ease future work (such as future edits) on the code for other people.

Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ❌ (This needs revision)
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ❌
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.✔
Unit test the Storage class using various entities to ensure that there are no bugs.✔
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.
Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: Logger/FullNameRecord.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+                 // Instead if we made these mutable objects, there is a chance we would overwrite something and "delete" a previous record.
+
+{
+
+    //here is the FullName struct, makes sure each full name has a first and last, middle can be null because not all 
+    //people may have a middle name
+    //struct allows for more freedom when using this for all of our other records\
+    // this is an implicit way of letting the compiler know that each full name always has a first and last name while allowing middle name to be null
+
+    public readonly record struct FullName(string First, string Last, string? Middle = null)
```
Comment (mmwoodfo) [inline Logger/FullNameRecord.cs:11]: These records should be in their own classes

File: Logger/FullNameRecord.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        public override string Name { get => Title; }
+
+    }
+
+
+    //this is the record that turns someones full name into a name, (First last middle) = name using our abstract class
+    public record Someone(FullName Full): AbstractEntity
+    {
+        public override string Name { get => Full.ToString(); }
```
Comment (mmwoodfo) [inline Logger/FullNameRecord.cs:36]: What do you expect ToString to print here? Consider adding a test for this.

File: Logger/AbstractEntity.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Logger
+{
+    public abstract record class AbstractEntity : IEntity
```
Comment (mmwoodfo) [inline Logger/AbstractEntity.cs:3]:
```suggestion
    public abstract record class BaseEntity : IEntity
```

File: Logger/IEntity.cs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Logger;
+public interface IEntity
+{
+
+    // Place members here.
+    public Guid Id { get; init; }
+    public string Name { get; }
```
Comment (mmwoodfo) [inline Logger/IEntity.cs:7]:
```suggestion
    Guid Id { get; init; }
    string Name { get; }
```

File: Logger.Tests/NameTests.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
+﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Logger.Tests
```
Comment (mmwoodfo) [inline Logger.Tests/NameTests.cs:3]: It looks like you are missing equality/inequality tests for your records

**Pull request closed without merging**